torch/testing/_internal/common_modules.py

```diff
@@ -85,9 +77,31 @@ def test_wrapper(*args, **kwargs):
                     for decorator in active_decorators:
                         test_wrapper = decorator(test_wrapper)
 
-                    yield (test_wrapper, test_name, param_kwargs)
+                    try:
+                        module_inputs = module_info.module_inputs_func(
+                            module_info, device=device_cls.device_type, dtype=dtype, requires_grad=False)
```

Comment:
I really like a lot of this, especially the "one test per input" idea, since it makes debugging really nice.

However, creating the module inputs at test discovery time may cause some problems. For example, consider a test that uses`@skipIfCUDA`. The decorator will be added at test discovery time, and the module input will also be instantiated on a (possibly non-existent) CUDA device also at test discovery time. So the decorator won't help skip the instantiation of a CUDA tensor and things will fail.

I wonder if we could still get "one test per input" somehow else - I guess we'd need the number of samples + names at test discovery time (without actual instantiation). That is, separate the metadata about the samples from the samples themselves.

Reply:
After going through several iterations of "one test per input", I think all of them complicates the developer API too much. The current way of generating samples when `module_inputs_func` is called is very simple to understand.

I updated the PR to just include new tests, and remove the adjustment to `common_modules`.

Reply:
Coming to this conversation pretty late: yes, you can't have one test per input unless build build a redundant "compile time" mechanism that has information about the samples that will be generated at "runtime." This is possible but very tricky.